pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java

```diff
@@ -299,16 +274,40 @@ private List<TableConfig> getListTableConfigs(List<String> tableNames) {
     return allTableConfigList;
   }
 
-  // return the brokerTenant if all table configs point to the same broker, else returns null
-  private String getCommonBrokerTenant(List<TableConfig> tableConfigList) {
-    Set<String> tableBrokers = new HashSet<>();
-    for (TableConfig tableConfig : tableConfigList) {
-      tableBrokers.add(tableConfig.getTenantConfig().getBroker());
+  private String selectRandomInstanceId(List<String> instanceIds) {
+    if (instanceIds.isEmpty()) {
+      return QueryException.BROKER_RESOURCE_MISSING_ERROR.toString();
+    }
+
+    instanceIds.retainAll(_pinotHelixResourceManager.getOnlineInstanceList());
+    if (instanceIds.isEmpty()) {
+      return QueryException.BROKER_INSTANCE_MISSING_ERROR.toString();
     }
-    if (tableBrokers.size() != 1) {
-      return null;
+
+    // Send query to a random broker.
+    return instanceIds.get(RANDOM.nextInt(instanceIds.size()));
+  }
+
+  private List<String> findCommonBrokerInstance(Set<String> brokerTenants) {
+    Set<String> commonInstances = null;
+    for (String brokerTenant : brokerTenants) {
+      Set<String> instances = _pinotHelixResourceManager.getAllInstancesForBrokerTenant(brokerTenant);
```

Comment:
It will be much faster if we first get all instance configs, and then pass in the instance configs when getting instance ids. See the `TODO` in `PinotHelixResourceManager.getAllInstancesForBrokerTenant()`